design/one-pager-component-mutable-and-versioning.md

```diff
@@ -0,0 +1,483 @@
+# Component mutable and versioning mechanism
+
+* Owner: Lei Zhang (@resouer), Jianbo Sun (@wonderflow)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+
+## Background
+
+OAM v1alpha1 made a requirement that developer need to define another ComponentSchematic with
+different name for an update (i.e. ComponentSchematic is immutable). We've got lots of feedback from
+real practices in Alibaba as well as [AWS ECS for OAM](https://github.com/awslabs/amazon-ecs-for-open-application-model#upgrade-and-scale-oam-workloads-with-oam-ecs)
+that it's not a good experience. Today [Component is mutable in v1alpha2 spec](https://github.com/oam-dev/spec/pull/356),
+this proposal is going to make it into implementation.
+
+Since component is becoming mutable, a very common question is how will the component versioning mechanism work.
+One of the requirement is the versioning mechanism should allow users to specify a fixed version of component
+so that the Application can be upgraded as before. On the other hand, the versioning mechanism should also be used
+by traits. For example, a trait can specify different revision and control deployment process. 
+
+
+## Goals
+
+1. Component is mutable. Developer who modify the Component object will directly trigger further
+upgrade of the application by default. 
+2. Component Version can be specified in ApplicationConfiguration. Operator can specify a fixed version
+of Component in ApplicationConfiguration. Component mutable won't affect Application directly in this case, 
+and can be upgraded in a more convenient way by working with trait.
+3. The versioning mechanism should compatible with current existing K8s workloads, eg, `K8s Deployment`, `OpenKruise`.
+  OAM should support two common ways used in community for versioning mechanism. 
+  - Workload itself can handle version. This means we can use `Deployment` only to do rollout or rollback without trait.
+  - Corporate with trait. OAM runtime should create different `Deployment` instances for different revision and let the trait do rollout or rollback.
+
+## Proposal
+
+### Using ControllerRevision to record versions of Component
+
+Because a `Component` is mutable, we introduce an independent Revision object to track a `Component`
+change automatically.
+
+We will use [`ControllerRevision`](https://godoc.org/k8s.io/api/apps/v1#ControllerRevision) in K8s as the Revision object.
```

Comment:
I like the idea of using existing k8s primitives here, but I worry that the `ControllerRevision` object is going to be a bit too lightweight for versioning of a `Component`. The `ControllerRevision` object is used by higher-level objects like `DaemonSet` and `StatefulSet` to handle update and rollback, but does not appear to be robust enough for a case like `Component` where an organization may want to "publish" multiple versions of a `Component` that will be consumed by multiple `ApplicationConfigurations`. For instance, if I was authoring a _new_ `ApplicationConfiguration`, specifying a `revisionName` that corresponds to an `int64` `revision` does not feel like a "application owner" level decision.

It seems like what we really want to capture here is versioning of an `ApplicationConfiguration`, for which the `ControllerRevision` feels much more appropriate. For versioning of a `Component`, I think I would actually advocate for making the current notion of a `Component` immutable, such that a developer would create a new `Component` entirely when they want it to be changed. However, this clearly makes it difficult to group `Components` together as they would have to have different names.

A solution to this would be to have a `VersionedComponent` and a `Component` (definitely open to different / better names :sweat_smile: ). The `VersionedComponent` would look very similar to the existing current implementation of a `Component`, but would include a `version` field and a `componentRef` field in the `spec`. The new `Component` would be a vehicle for keeping track of `VersionedComponents`.

Let's explore some tangible examples of what this could look like:
1. Create a `Component` called `frontend`
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: Component
metadata:
  name: frontend
spec:
  # versions will start out as empty
```
2. Create the first version of the `frontend` as a `VersionedComponent`
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: VersionedComponent
metadata:
  name: frontend-v0.0.1 # you could name this anything you wanted, but just appending the semver would be good practice
spec:
  version: v0.0.1
  componentRef:
    name: frontend
  workload:
    apiVersion: core.oam.dev/v1alpha2
    kind: ContainerizedWorkload
    spec:
      containers:
      - name: my-cool-workload
        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
        cmd:
        - "bash lscpu"
```
3. A `Component` controller sees this and adds an entry to the `Component` with the `VersionedComponent`:
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: Component
metadata:
  name: frontend
spec:
  versions:
    - v0.0.1: frontend-v0.0.1
```
4a. An `ApplicationConfiguration` is created that references the `Component` and a `version`:
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend
      version: v0.0.1 # this would be optional
```
4b. An `ApplicationConfiguration` is created that references a `Component` only:
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend # This would result in the latest semantic version being used from the versions map on the Component
```

Now let's look at update strategy. We could do as suggested in the proposal in this PR and if no version is specified then we automatically bump to new versions added to a `Component`. This would be handled by the `ApplicationConfiguration` controller. On each reconcile it would `GET` the `Component` and check to make sure that the latest version is being used. If not, then it will update the rendered `ContainerizedWorkload` in this case. This is where the `ControllerRevision` could come in, if we either automatically update due to a new version being added to the `Component` or manually update due to an explicit `version` change on the `ApplicationConfiguration`, we can create a `ControllerRevision` (or some other new object type if necessary) to handle update and rollback. This makes more sense in my opinion because we are not interested in update and rollback of components themselves, but rather of the _usage_ of components in an `ApplicationConfiguration`.

This would stil allow for usage of traits such as `Rollout` or `TrafficSplit`. but they would only need to specify semantic versions instead of specific revision names.

Reply:
Hi Daniel, thanks for detailed review comment. Yes, we agree with you creating a new revision object is neat since k8s' `ControllerRevision` is a generic purpose snapshot object with `data` section, it's not ideal for a user facing revision recorder. I think @wonderflow is more like trying to minimize change to the spec since we don't have a structured way to promote spec change.

So here's my suggestion: 
-  we propose a new revision object in this proposal per your suggestion (need to pick a name: ComponentRevision, VersionedComponent etc);
   - consider: `apiVersion: experimental.oam.dev/v1alpha2`.
- I will try to work out a way to enable implementation driven and versioned spec change promotion mechanism, so we can introduce change safely to spec. 
    - the idea is we will add `experimental` to api version of new object (see above) and new field, see https://github.com/crossplane/oam-kubernetes-runtime/pull/24#issuecomment-634840881. And we will cut release in OAM spec by promoting them to `core.oam.dev/v1alpha3` and remove the `experimental` tag.

Would be glad to know your ideas on this part as well.

Snapshotting ApplicationConfiguration is not in scope for this PR, but I agree we will use ControllerRevision in this case.

Reply:
Yeah, I'm agree with using `ComponentRevision` instead of `ControllerRevision`. But for this:

```
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: frontend
      version: v0.0.1 # this would be optional
```

I think we can still use `revisionName` to do so, since we have [discussed a lot](https://github.com/oam-dev/spec/issues/350)  before and have already [merged in spec](https://github.com/oam-dev/spec/pull/356). They are almost the same.